myia/grad.py

```diff
@@ -0,0 +1,284 @@
+
+from typing import Set
+
+from .anf_ir import Apply, Constant, Parameter, Graph, ANFNode
+from . import primops
+from collections import defaultdict
+from .anf_ir_utils import dfs
+from .cconv import NestingAnalyzer
+
+
+def is_constant(x):
+    return isinstance(x, Constant)
+
+
+def is_graph(x):
+    return isinstance(x, Constant) and isinstance(x.value, Graph)
+
+
+add = Constant(primops.Add())
+J = Constant(primops.J())
```

Comment:
I might have asked this before, but I can't remember the answer: Why do we want to have `J` as a primitive? Why not "evaluate" `J` directly at parse-time here? It only seems to be applied to constants after all.

Reply:
`J` will not always be applied to constants. It will also be applied to e.g. arguments to higher order functions, in which case we need to do some value propagation. We might need a runtime version of `J` that looks up the proper pre-transformed function in a static map.

Also, it's quite a bit simpler to implement and debug a transformation if you don't have to worry about recursively expanding macros. Many operations we wish to eliminate can exist in the IR transiently, I don't think this is a problem.

Reply:
Do you have a case in which a runtime `J` primitive is required? If you want to keep `J` around for one pass and remove it another, sure, but I'm wondering whether we can get rid of `J` before runtime/codegen, in which case we probably should. Targeting different backends would be easier if we don't have AD-specific primitives floating around.

Reply:
I don't know, and I'm not too concerned about that for now. The Python VM can implement that primitive to run the transformer, which could be nice for debugging, and the same system if tasked with constant prop can take care of eliminating it. As for a potential C or Rust VM, it would basically be just a lookup, but I think for now we can optimistically assume we won't need it.